### ISSUE TITLE
Sync client removing data from server when OS drive is full

### ISSUE THREAD

**Issue by zacharlie** [nextcloud/desktop#3076]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

---

### Behaviour

When the primary disk on the operating system reaches capacity, the NextCloud desktop sync client starts deleting files on the server (and removed from other client devices utilising the same shared directories).

This was first outlined in #2831  and discussed further in #1433 

### Steps to reproduce

It seems that in each of the reported cases, this occurs when users on Windows have their sync directory on another drive.

### Additional notes

The deleted data was available from my "deleted files" directory on the web client, however the server ran out of storage space (unrelated reasons) whilst attempting to restore data and the process was interrupted, so another restoration procedure was initiated instead.

It is unknown if the data restoration from the users deleted files would have been successful.

### Client configuration
Client version: 3.1.1

Operating system: Windows 10 Pro, Version 20H2 Build 19042

Installation path of client: `C:\Program Files\Nextcloud`

Folder sync location: `P:\nextcloud`

### Logs

I will not publicly share my log data at this time for data privacy reasons

**Comment by zacharlie** [comment]
@FlexW could you unstale this issue? I think it's a pretty critical one...

**Comment by mgallien** [comment]
@zacharlie I have reopen it but could you please test with the latest version to see if this is still happening (would be really nice as I am not sure I will be able to reproduce) ?

I will put the needs info label and without answer in 6 weeks it will get closed again.

**Comment by zacharlie** [comment]
I won't be testing on our production instance (as it is dangerous) and have limited resources, however I will try to see if I can get it set up with a test instance in similar configuration and reproduce/ test using what I have available. If anyone else out there can test as well it would be appreciated.

**Comment by PVince81** [comment]
This happened to me as well yesterday with nextcloud-desktop 3.1.3 on openSUSE Leap 15.3.

My partitions look like this:

- "/" with btrfs on a SSD
- "/home" with ext4 on a HDD with plenty of space
- "/home/vincent/{.cache,.local}" are symlinked to the SSD for performance reasons
- multiple sync folders configured but the biggest one is "/home/vincent/cloud/public/Photos" with subdirs by year 1980, 2000, 2002, ... up to 2021 and one top level folder "Family". In whole it's about 1 TB of data.

The 4 other sync folders were apparently untouched by the incident, probably because the client was busy for long enough with that one huge one.

At around 19:55 CEST yesterday, the client started to DELETE all subfolders of "Photos", this is a piece of the sync log for the configured "Photos" sync folder:
```
#=#=#=# Syncrun started 2021-06-07T19:14:09Z
#=#=#=#=# Propagation starts 2021-06-07T19:14:16Z (last step: 6433 msec, total: 6433 msec)
#=#=#=# Syncrun finished 2021-06-07T19:14:16Z (last step: 511 msec, total: 6945 msec)
#=#=#=# Syncrun started 2021-06-08T15:54:18Z
#=#=#=#=# Propagation starts 2021-06-08T15:54:56Z (last step: 38308 msec, total: 38308 msec)
#=#=#=# Syncrun finished 2021-06-08T15:54:57Z (last step: 507 msec, total: 38816 msec)
#=#=#=# Syncrun started 2021-06-08T17:55:02Z
#=#=#=#=# Propagation starts 2021-06-08T17:55:25Z (last step: 22192 msec, total: 22192 msec)
#=#=#=# Syncrun finished 2021-06-08T17:55:25Z (last step: 473 msec, total: 22665 msec)
#=#=#=# Syncrun started 2021-06-08T17:55:32Z
#=#=#=#=# Propagation starts 2021-06-08T17:55:38Z (last step: 5835 msec, total: 5835 msec)
17:55:38||2018|INST_REMOVE|Up|1607780457|5fd4c87fb9d19|0|00118540ocu6qxzdhx2s|4||204|0|0||||
17:55:38||Family|INST_REMOVE|Up|1622727440|60b8db10967cc|0|00118534ocu6qxzdhx2s|4|Local files and share folder removed.|204|0|0||||
17:55:38||2021|INST_REMOVE|Up|1623085939|60be53730a16b|0|00329289ocu6qxzdhx2s|4||204|0|0||||
17:55:50||2017|INST_REMOVE|Up|1607780882|5fd4ca26b5108|0|00118564ocu6qxzdhx2s|4||204|0|0||||
17:55:55||2015|INST_REMOVE|Up|1607780583|5fd4c91a4dfa8|0|00118548ocu6qxzdhx2s|4||204|0|0||||
17:55:53||2016|INST_REMOVE|Up|1607780566|5fd4c8dd33674|0|00118542ocu6qxzdhx2s|4||204|0|0||||
17:55:58||2014|INST_REMOVE|Up|1607780627|5fd4c91a4dfa8|0|00118551ocu6qxzdhx2s|4||204|0|0||||
17:56:06||2012|INST_REMOVE|Up|1607780385|5fd4c848e6071|0|00118538ocu6qxzdhx2s|4||204|0|0||||
17:56:06||2011|INST_REMOVE|Up|1607780838|5fd4ca26b5108|0|00118562ocu6qxzdhx2s|4||204|0|0||||
17:56:14||2009/2009-12-26 AAAAAAAA|INST_REMOVE|Up|1607780370|5fd4c848e6071|0|00135175ocu6qxzdhx2s|4||204|0|0||||
17:56:05||2013|INST_REMOVE|Up|1607780275|5fd4c7d597bb0|0|00118530ocu6qxzdhx2s|4||204|0|0||||
17:56:12||2010|INST_REMOVE|Up|1607780609|5fd4c91a4dfa8|0|00118549ocu6qxzdhx2s|4||204|0|0||||
17:56:18||2009/2009-12-23 NNNNNM|INST_REMOVE|Up|1607780350|5fd4c80bafca5|0|00135149ocu6qxzdhx2s|4||204|0|0||||
17:56:18||2009/2009-12-24 MMMMMM|INST_REMOVE|Up|1607780372|5fd4c848e6071|0|00135192ocu6qxzdhx2s|4||204|0|0||||
17:56:18||2009/2009-12-23 YYYYY|INST_REMOVE|Up|1607780349|5fd4c80bafca5|0|00135142ocu6qxzdhx2s|4||204|0|0||||
17:56:22||2009/2009-12-21 ZZZZ|INST_REMOVE|Up|1607780358|5fd4c80bafca5|0|00135157ocu6qxzdhx2s|4||204|0|0||||
```

Note: the log is in UTC so 17:55 matches 19:55 CEST.
I've replaced some names there with letters.

It's only around 23:15 that I noticed the sync icon staying so I went to look and saw that it was reuploading everything into the Photos folder. Then around 23:30 my desktop crashed, this is how I noticed that the root partition was full.
I can't pinpoint the exact moment where the disk was "too full".

Looking at the log, the weird thing is that it didn't remove all the top level directories. It removed a few years, but for the 2009 folder it only removed stuff inside but not the folder itself.

Much later after the deletion wave, it started reuploading again, here's a snippet between the last deletions and the beginning of the reuploads:
```
18:05:47||2009/2009-06-09 上海/dsc04071.jpg|INST_REMOVE|Up|1320583030|99a64c96ee1593cab2c85e5aace9bdbc|782773|00137193ocu6qxzdhx2s|4||204|0|0||||
18:05:47||2009/2009-06-09 上海/dsc04072.jpg|INST_REMOVE|Up|1320583030|541ae8b9cd1d4b455ab31bc7c773f6ad|827583|00137258ocu6qxzdhx2s|4||204|0|0||||
18:05:47||2009/2009-08-30 iiiii.jpg|INST_REMOVE|Up|1320581338|0555078cfe5f8473c7298b9f9f189a13|82608|00135139ocu6qxzdhx2s|4||204|0|0||||
18:05:47||common.txt|INST_REMOVE|Up|1608474228|2ee108e9379c585d82e52cf91ef7b494|320|00244727ocu6qxzdhx2s|4||204|0|0||||
18:05:47||kdenlive-sources.txt|INST_REMOVE|Up|1608474312|2c6435caf03cac36d10bc971c9626ffe|1280|00244719ocu6qxzdhx2s|4||204|0|0||||
18:05:47||videos-bigger-400m.txt|INST_REMOVE|Up|1608474347|c30e4c8e6a066e78e7af7d4964787801|7980|00244463ocu6qxzdhx2s|4||204|0|0||||
#=#=#=# Syncrun finished 2021-06-08T18:05:55Z (last step: 616666 msec, total: 622501 msec)
#=#=#=# Syncrun started 2021-06-08T18:06:35Z
#=#=#=#=# Propagation starts 2021-06-08T18:06:40Z (last step: 4499 msec, total: 4499 msec)
#=#=#=# Syncrun finished 2021-06-08T18:06:40Z (last step: 220 msec, total: 4719 msec)
#=#=#=# Syncrun started 2021-06-08T20:06:42Z
#=#=#=#=# Propagation starts 2021-06-08T20:07:25Z (last step: 42860 msec, total: 42860 msec)
20:07:26||198X|INST_NEW|Up|1604352628||4096|00495291ocu6qxzdhx2s|4||201|0|0||||
20:07:26||1987|INST_NEW|Up|1608056223||4096|00495292ocu6qxzdhx2s|4||201|0|0||||
20:07:26||1984|INST_NEW|Up|1604348145||4096|00495290ocu6qxzdhx2s|4||201|0|0||||
20:07:26||1981|INST_NEW|Up|1604352542||4096|00495293ocu6qxzdhx2s|4||201|0|0||||
20:07:26||1980|INST_NEW|Up|1604348153||4096|00495294ocu6qxzdhx2s|4||201|0|0||||
20:07:26||1992|INST_NEW|Up|1604346911||4096|00495295ocu6qxzdhx2s|4||201|0|0||||
20:07:26||1984/1984 CCCCCC|INST_NEW|Up|1364731352||4096|00495296ocu6qxzdhx2s|4||201|0|0||||
20:07:26||1980/1980 YYYYY|INST_NEW|Up|1364740119||4096|00495297ocu6qxzdhx2s|4||201|0|0||||
20:07:26||1980/1980 VVVVV|INST_NEW|Up|1364740526||4096|00495298ocu6qxzdhx2s|4||201|0|0||||
```

Now, the question is why it started to delete in the first place ?

Does the system call that checks for folder existence fail (due to full root partition) but the client interprets this as "does not exist" instead of failing ? Would be good to check that first.

**Comment by PVince81** [comment]
Also interesting is that before the mass deletion, it did a successful propagation test with no syncing.

And after the big deletion, it also has one cycle where it discovers and finds nothing to change. So it means there are two points in time where the desktop client found that the folders were "missing" in some way and thought "all is fine".

And only after the second retry suddenly the folders were found again.

Where's the code snippet that checks for folder existence ?

**Comment by pierreozoux** [comment]
It is really scary, I'd recommend people to 👍

**Comment by mgallien** [comment]
> It is really scary, I'd recommend people to +1

Please keep in mind that this ticket is probably a duplicate of #1433 .

Unfortunately no maintainers managed to reproduce it so if you still encounter it, please send us the logs. That also means that full disc is not the trigger but some other particular conditions. That also means that it has a very low probability of occurrences.

A lot of thought, tests and modifications went into the last releases (3.2.2 in particular) to try to fix that bug once and for all. We are still actively working on it.

**Comment by maxxer** [comment]
> Unfortunately no maintainers managed to reproduce it so if you still encounter it, please send us the logs. That also means that full disc is not the trigger but some other particular conditions. That also means that it has a very low probability of occurrences.

We run into this quite regularly, unfortunately. Our users manage lots of data, so they have the Nextcloud data dir in D: while Nextcloud stores its internal database in C:. When C fills up there's no check on the NC side so I believe the db gets corrupted and the issue is triggered.

I know this sync db used to be in the same partition of the NC data dir at some point, and then was moved to system drive recently. Maybe having it on the same disk as data was a better choice to both prevent fillup and corruption, but if I recall there were other issues.
